[WIP] Vfio interrupt support

tzwickl opened the pull request:
Add VFIO interrupt support for ixy driver.

tzwickl committed:
Add first version of interrupts.

tzwickl committed:
Add epoll and eventfd support.

tzwickl committed:
Cleanup interrupts.

tzwickl committed:
Change comments to rust style.

tzwickl committed:
Remove eventfd as it is outdated.

tzwickl committed:
Code cleanup.

tzwickl committed:
Fix compilation errors.

tzwickl committed:
Fix compilation errors.

tzwickl committed:
Fix error handling.

tzwickl committed:
Fix error handling.

tzwickl committed:
Revert "Fix error handling."

This reverts commit 3ab612460630fb6fff6dc353ecf899b98df7fa1d.

tzwickl committed:
Fix errors.

tzwickl committed:
Fix error.

tzwickl committed:
Fix error.

tzwickl committed:
Fix error.

tzwickl committed:
Fix warnings.

tzwickl committed:
Remove unused code.

tzwickl committed:
Fix some warnings.

tzwickl committed:
Add missing variables.

tzwickl committed:
Fix errors.

tzwickl committed:
Fix various bugs.

tzwickl committed:
Fix some more bugs.

tzwickl committed:
Move ITR to IXGBE driver.

tzwickl committed:
Improve moving average.

ackxolotl commented:
Is this PR ready to merge or still work in progress? If the former, please rebase your commits to master to resolve the merge conflicts. If the latter, do you want me to review your code now?

tzwickl commented:
> Is this PR ready to merge or still work in progress? If the former, please rebase your commits to master to resolve the merge conflicts. If the latter, do you want me to review your code now?

Hey :)
It's not ready to merge yet, but would be nice if you could still review it and give me some feedback. Afterward, I will implement your feedback, resolve the conflicts and re-base it to master :)

ackxolotl commented on src/interrupts.rs:
```diff
@@ -0,0 +1,305 @@
+use epoll::Event;
+use epoll::Events;
+use std::time::Instant;
+use libc;
```
`use libc` can be removed

ackxolotl commented on src/ixgbe.rs:
```diff
@@ -787,6 +832,188 @@ impl IxgbeDevice {
+
+        // Step 3: All interrupts should be set to 0b (no auto clear in the EIAC register). Following an
+        // interrupt, software might read the EICR register to check for the interrupt causes.
+        self.set_reg32(IXGBE_EIAC, 0x00000000);
```
Guess you could write `0x0` or simply `0` here ;) Or write `0x0000_0000` for better readability if you want to make this explicit

ackxolotl commented on src/ixgbe.rs:
```diff
@@ -787,6 +832,188 @@ impl IxgbeDevice {
+        self.set_ivar(0, queue_id, 0);
+
+        // Step 2: Program SRRCTL[n].RDMTS (per receive queue) if software uses the receive
+        // descriptor minimum threshold interrupt
```
So there is nothing done here ... can you insert a one line comment stating why? The same would be nice for step 4 ... otherwise it looks like you forgot something to do there ^^

ackxolotl commented on src/ixgbe.rs:
```diff
@@ -274,6 +294,26 @@ impl IxyDevice for IxgbeDevice {
+                interrupt.rx_pkts += received_packets as u64;
+
+                let elapsed = interrupt.last_time_checked.elapsed();
+                let diff = elapsed.as_secs() * 1_000_000_000 + elapsed.subsec_nanos() as u64;
```
Please adapt this like you did with the C version :)

ackxolotl commented on src/interrupts.rs:
```diff
@@ -0,0 +1,305 @@
+const VFIO_DEVICE_SET_IRQS: u64 = 15214;
+pub const VFIO_PCI_MSI_IRQ_INDEX: u64 = 1;
+pub const VFIO_PCI_MSIX_IRQ_INDEX: u64 = 2;
+const VFIO_IRQ_INFO_EVENTFD: u32 = (1 << 0);
```
These structs and constants might fit better into `vfio.rs` to have all vfio related structs and constants in one file?

Btw: This is one of the reasons you should rebase :D There is no `vfio.rs` in your branch yet

ackxolotl commented on src/lib.rs:
```diff
@@ -13,10 +13,12 @@ mod constants;
 mod ixgbe;
 pub mod memory;
 mod pci;
+pub mod interrupts;
```
Should this module be public? And do the constants, structs and methods in the module have to be public? If not, please make them private or `pub(crate)`

ackxolotl commented on src/interrupts.rs:
```diff
@@ -0,0 +1,305 @@
+    }
+
+    /// Waits for events on the epoll instance referred to by the file descriptor `epoll_fd`.
+    /// The memory area pointed to by events will contain the events that will be available for the caller.
```
Please insert a new line after the head line, e.g.

```rust
/// x does y.
///
/// This can be used to ...
fn x() {}
```

All in all, please adjust your doc comments to be more consistent with existing comments (only doc comments, the other comments are good to go)

ackxolotl commented on src/ixgbe.rs:
```diff
@@ -395,18 +435,18 @@ impl IxyDevice for IxgbeDevice {
 
-        // section 4.6.3.1 - disable interrupts again after reset
-        self.set_reg32(IXGBE_EIMC, 0x7fff_ffff);
+        // section 4.6.3.1 - disable all interrupts
```
Maybe add `(disable again after reset)` here?

ackxolotl commented on Cargo.toml:
```diff
@@ -8,6 +8,7 @@ edition = "2018"
 libc = "0.2"
 byteorder = "1"
 log = "0.4"
+epoll = "4.1.0"
```
Maybe we can waive this crate? :) Doesn't do a lot (only one file with less than 150 lines of code) but it includes yet another crate in our driver :-/

ackxolotl reviewed (changes requested):
So all in all this looks quite good yet :)

For some more comments from my side you should
* add some more logging information, the C version has stuff like `info("Using MSIX interrupts");` in it, this is missing in Rust (use `info!("something");` or `debug!("something");` where appropriate)
* run `cargo clippy` and fix everything that sounds reasonable to you
* run `cargo fmt` (if there is anything you want to have formatted in another way, add a `#[rustfmt::skip]`) before that line
* rebase your code on master

tzwickl committed:
Implement feedback from code review.

tzwickl committed:
Remove epoll crate.

tzwickl committed:
Merge branch 'master' into vfio-interrupt

tzwickl committed:
Finalize merge.

tzwickl committed:
Fix some minor things.

tzwickl committed:
Format code.

tzwickl committed:
Fix some calls.

tzwickl committed:
Make OS errors safe.

tzwickl committed:
Format code.

ackxolotl committed:
Assimilate vfio.rs

We can replace `pub(crate)` by `pub` since the module isn't public anymore, makes `vfio.rs` more consistent

ackxolotl committed:
Update interrupts.rs

ackxolotl committed:
Update interrupts.rs

ackxolotl committed:
Update interrupts.rs

ackxolotl committed:
Update echoer.rs

ackxolotl committed:
Update forwarder.rs

ackxolotl committed:
Update generator.rs

ackxolotl committed:
Update interrupts.rs

tzwickl committed:
Add missing flags.

ackxolotl commented on src/interrupts.rs:
```diff
@@ -0,0 +1,330 @@
+
+    /// Enable VFIO MSI interrupts for the given `device_fd`.
+    pub fn vfio_enable_msi(&mut self, device_fd: RawFd) -> Result<(), Box<dyn Error>> {
+        info!("Enable MSI Interrupts");
```
Please adapt your style to match the other `info!()` macro calls, it should read `info!("enabling MSI interrupts")`

ackxolotl replied:
This also applies to other `info!("")` and `Err("".into())` ;)

ackxolotl reviewed (changes requested):
Comments and error messages should still be adapted to the existing style and then I'm going to squash and merge this :)

tzwickl committed:
Fix undefined behavior in VFIO module

tzwickl committed:
Adjust errors and info messages

tzwickl committed:
Merge branch 'master' into vfio-interrupt

tzwickl committed:
Add some more logging.

ackxolotl committed:
Merge branch 'master' into vfio-interrupt

ackxolotl committed:
Update pcap.rs

ackxolotl committed:
Update lib.rs

ackxolotl committed:
Update virtio.rs

ackxolotl commented:
@ixy-ci test
(just for fun, this is virtio only)

ackxolotl commented:
@ixy-ci ping

tzwickl committed:
Update threshold.

tzwickl committed:
Adjust interrupt threshold.

ackxolotl committed:
Fix indentation

ackxolotl commented on src/interrupts.rs:
```diff
@@ -0,0 +1,328 @@
+            )
+            .into());
+        }
+        let epoll_fd = status as RawFd;
```
Please use `epoll_fd` directly instead of `status`, i.e. `let epoll_fd: RawFd = unsafe { libc::epoll_create1(0) };`

tzwickl replied:
Same here => `rc` is `usize` so how can I check later in the code if `status == -1` to check for errors?

ackxolotl replied:
Which `rc`?

tzwickl replied:
ahh I meant `epoll_fd` :)

ackxolotl replied:
`RawFd` is `c_int` which will almost always be a `i32`. So you can check if it is `-1` :)

tzwickl replied:
ahh right my bad yes :) will change it :)

ackxolotl commented on src/interrupts.rs:
```diff
@@ -0,0 +1,328 @@
+            )
+            .into());
+        }
+        rc = status as usize;
```
Replace `status` by `rc` (same as above)

tzwickl replied:
This does not work right? Because `rc` is `usize` so how can I check later in the code if `status == -1` to check for errors?

ackxolotl replied:
Right, my bad. But `vfio_epoll_ctl()` should be fixable.

tzwickl replied:
it's the same for `vfio_epoll_ctl`

ackxolotl replied:
No? :D

ackxolotl commented on src/interrupts.rs:
```diff
@@ -0,0 +1,328 @@
+                    libc::read(
+                        event.data as i32,
+                        val_ptr as *mut libc::c_void,
+                        mem::size_of::<u64>(),
```
Either this should read `mem::size_of::<u16>()` or `val` should be `u64` ... and `val_ptr` is a bit redundant, just do a `&mut val as *mut _ as *mut libc::c_void` here

tzwickl committed:
Simplify code.

tzwickl commented:
@ackxolotl should be fixed now :)

ackxolotl commented on src/interrupts.rs:
```diff
@@ -0,0 +1,326 @@
+        if rc > 0 {
+            /* epoll_wait has at least one fd ready to read */
+            for event in events.iter().take(rc) {
+                let mut val: u16 = 0;
```
[In the C code](https://github.com/emmericp/ixy/blob/master/src/libixy-vfio.c#L191), `val` is a `uint64_t`. What is correct?

tzwickl replied:
Does not matter ;) you specify the size it should read anyway so it reads either `u16` or a `u64`. We just do this to reset the interrupt in the eventfd file so we don't care about how much we read but just that we read something :)

tzwickl replied:
I changed it to `u64` so it is consistent with the C code :)

ackxolotl reviewed (commented):
See comment

tzwickl committed:
Make code consistent with C.
